[cmd/serverless] add support for Logs Collection in the Datadog Agent Extension

remeh opened the pull request:
### What does this PR do?

This PR adds support for Logs collection in the Datadog Agent Extension.

#### Log collection

In order to collect the logs from the environment, the extension is adding a route `/lambda/logs` into the extension HTTP server (running on port 8124) on which it is receiving logs from the AWS environment, these logs are the logs of the function, of the extensions and of the platform.

The `pkg/serverless/aws` package has been introduced with methods helping to get the ARN of the running function + the parsing of the log message received on the HTTP server.

#### Logs aggregation / processing

An instance of the Logs Agent is started in the extension if the `DD_LOGS_ENABLED` flag is set. This Logs Agent use the regular pipeline except that:
  - it is using a newly introduced [`NullAuditor`](https://github.com/DataDog/datadog-agent/pull/6861/files#diff-a9e8b62ad98aa2d1b07a4fa7ceb37a3e55b456d7a5ecf0044cf7c2878cadf54dR7-R13) (not writing any registry).
  - a new [input source](https://github.com/DataDog/datadog-agent/pull/6861/files#diff-e850beb91ee56536bd782a5ffa1ba6f02da879714b0491861f1891f5b99fa728R43) has been introduced capable of receiving logs message from AWS in a channel (see `pkg/logs/input/channel`). Note that it is specialized for AWS logs message but could be more generic with additional code if needed.
  - The JSON encoder has been [modified](https://github.com/DataDog/datadog-agent/pull/6861/files#diff-40507f3cd5a5844dd542647018eebd84a22d6fa6fb19736ae758b643edb2ede4R36) to support providing a timestamp to messages, and, using the build tag `serverless`, it is not appending any hostname to the logs.

#### Synchronous flush of the pipeline

The Extension needs to do a synchronous flush of the buffered data on a signal. Today, it is used to naively flush at the end of the execution of the function, however, we already work on adding a smarter flush mechanism.

In order to do so, I had to adapt different parts of the pipeline:
  - Add `Pipeline.Flush`, triggering a flush of the Processor and of the Sender.
  - Add `Processor.Flush`, sending buffered message to the processor / encoder
  - Add `Sender.Flush`, sending the buffered data to the BatchStrategy sender
  - Add `BatchStrategy.Flush` to synchronously send the HTTP request to the intake to flush the data.

A mutex has been introduced and the `select` were all modified to support a signal for the flush. This may have some performances impact for which I want to add benchmarks.

#### Miscellaneous changes

  - Add [`message.NewMessageWithTime`](https://github.com/DataDog/datadog-agent/pull/6861/files#diff-7cc6d213f0f04ed2a60509dcaf928f38ec659a23de87d1386a624cae4bce2e2cR38) to send in the pipeline a log with a known/given timestamp
  - `Auditor` has been renamed `RegistryAuditor` and an [interface `Auditor`](https://github.com/DataDog/datadog-agent/pull/6861/files#diff-12a608d15229e66819669cdd4d22a67ddf843b7edeebeb7c204bfd1a11fd0f7dR55) has been introduced instead (to create the `NullAuditor`)
  - Some of the message contains useful information that we turn into metrics and directly into the aggregator: [see here](https://github.com/DataDog/datadog-agent/pull/6861/files#diff-298c068fdf582d7d88a0968c66fdbf83a9a7fb335d5537e5f550fc36362d54aaR103).

remeh committed:
Add the Serverless Datadog Agent implementation.

See https://docs.datadoghq.com/serverless/datadog_lambda_library/extension/

remeh committed:
Merge remote-tracking branch 'origin/master' into remeh/serverless

remeh committed:
Fix 3rd party license list since new pkg are imported from the AWS SDK.

remeh committed:
forwarder: adapt the SyncDefaultForwarder to recent changes in the Forwarder interface.

remeh committed:
aggregator: fix unit tests with Flush now being a public function.

remeh committed:
serverless: linter compliance.

remeh committed:
serverless: errcheck compliance.

remeh committed:
serverless: unused code linter.

remeh committed:
serverless: linter compliance.

remeh committed:
serverless: add http server listening for server logs and sending logs to the logs agent

Introducing the `logs/input/channel` implementation supporting receiving logs through
a Go channel.

remeh committed:
serverless/logs: introduce a `NullAuditor` not doing anything with messages.

This has been introduced to not change the original pipeline process.
The original `Auditor` has been renamed `RegistryAuditor`.
`Auditor` is now an interface for both `RegistryAuditor` and `NullAuditor`.

remeh committed:
serverless: remove debug messages

remeh committed:
serverless/logs: add a sync `Flush()` method to the logs pipeline.

This commit is also adding the reading of the function ARN from the invoke event.

remeh committed:
serverless/logs: remove unwanted debug lines.

remeh committed:
serverless/logs: read REPORT messages and send AWS enhanced metrics using the DogStatsD server.

Currently supported enhanced metrics: billed_duration, duration, init_duration, max_memory_used.

remeh committed:
serverless/logs: enhanced metrics are distribution.

remeh committed:
serverless/logs: comment on the sync flush of logs then statsd.

remeh committed:
serverless/logs: configurable logs type we're subscribing to.

This commit also adds the parsing of the time in the AWS log.

remeh committed:
serverless/arn: remove version from the ARN string if any.

remeh committed:
serverless/logs: create aws package containing both AWS LogMessage and ARN support.

This way, we can use the timestamp from the AWS LogMesage while sending the logs
to the intake. This is what this commit is doing.

remeh committed:
serverless/logs: parse time up to the millisecond.

remeh committed:
serverless/logs: send the enhanced metrics with the proper timestamp.

remeh committed:
serverless/logs: comments and send the report log to the intake.

remeh committed:
serverless/logs: send aws.lambda.enhanced.memorysize

remeh committed:
serverless/logs: send formated platform logs.

remeh committed:
serverless/logs: cleaning comments here and there.

remeh committed:
Merge branch 'master' into remeh/serverless

remeh committed:
config: fix log conflict while adding the serverless format.

remeh committed:
serverless/logs: linter compliance.

remeh committed:
Merge branch 'master' into remeh/serverless-logs

remeh committed:
Merge branch 'master' into remeh/serverless

remeh committed:
cmd/serverless: start -> run

remeh committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into remeh/serverless

remeh committed:
dogstatsd: add an unit test in serverless mode.

remeh committed:
tasks: remove serverless test tag

remeh committed:
Merge branch 'master' into remeh/serverless

remeh committed:
Merge branch 'master' into remeh/serverless-logs

remeh committed:
forwarder: rename `SyncDefaultForwarder` to `SyncForwarder` since it's not a default forwarder.

remeh committed:
Merge branch 'remeh/serverless' into remeh/serverless-logs

remeh committed:
Merge branch 'master' into remeh/serverless-logs

remeh committed:
[cmd/serverless] change default port for the HTTP server collecting logs

remeh committed:
serverless: support using DD_TAGS to add extra tags while sending metrics and logs.

For feature-parity with the Datadog Forwarder, we want to use this
configuration entry / environment variable to append tags to every
metric and log sent by the extension.

gh123man commented on pkg/logs/input/channel/tailer.go:
```diff
@@ -0,0 +1,59 @@
+
+	// Loop terminates when the channel is closed.
+	for logline := range t.inputChan {
+		origin := message.NewOrigin(t.source)
```
I think we should also update `t.source.BytesRead` to be consistent with the other tailers.

remeh committed:
serverless/logs: send the logs with the ARN and RequestId information.

remeh committed:
serverless: linter compliance

prognant commented on pkg/serverless/protocol.go:
```diff
@@ -56,6 +76,98 @@ func StartDaemon(stopCh chan struct{}) *Daemon {
 
+// StartHTTPLogsServer starts an HTTP server, receiving logs from the AWS platform.
+// Returns the HTTP URL on which AWS should send the logs.
+// FIXME(remy): that would be awesome to have this directly running within the initial HTTP daemon?
```
I guess that would make sense to put everything under `/lambda/...` and re-use the existing HTTP server.

remeh replied:
Finally took the time to address this 👍

prognant reviewed (commented):
Overall it looks good to me and fits pretty well within the existing Logs Agent code.

remeh commented on pkg/logs/auditor/null_auditor.go:
```diff
@@ -0,0 +1,55 @@
+// NullAuditor is an auditor not doing anything with the message it received.
+// It has been introduced for the Serverless Agent which doesn't need
+// to maintain a registry.
+type NullAuditor struct {
+	channel     chan *message.Message
+	stopChannel chan struct{}
+}
```
Interesting, I'll look into this

remeh replied:
That's indeed a solution, however, the original implementation of the Auditor is trying to do stuff with the registry on disk (for instance trying to read/recovery it at startup) which we don't want to do in this environment. Because of this, this `NullAuditor` is slightly more broader than just doing no-op on every message.
I think I'll keep the `NullAuditor` because of that and indeed, we may consider using it for some other tailers implementation if that makes sense.

prognant replied:
I agree having no operation on-disk is much more relevant here.

remeh commented on pkg/logs/pipeline/provider.go:
```diff
@@ -20,30 +20,44 @@ type Provider interface {
 type provider struct {
 	numberOfPipelines int
-	auditor           *auditor.Auditor
+	auditor           auditor.Auditor
```
It is because `auditor.Auditor` is now an interface, and not anymore a concrete type;

remeh commented on pkg/serverless/serverless.go:
```diff
@@ -6,17 +6,29 @@ import (
-	routeRegister  string = "http://localhost:9001/2020-01-01/extension/register"
-	routeEventNext string = "http://localhost:9001/2020-01-01/extension/event/next"
-	routeInitError string = "http://localhost:9001/2020-01-01/extension/init/error"
+	routeRegister      string = "http://localhost:9001/2020-01-01/extension/register"
+	routeEventNext     string = "http://localhost:9001/2020-01-01/extension/event/next"
+	routeInitError     string = "http://localhost:9001/2020-01-01/extension/init/error"
+	routeSubscribeLogs string = "http://localhost:9001/2020-08-15/logs"
```
Let me try this out

remeh committed:
serverless/logs: use the function name if available as the "service".

remeh committed:
serverless: hit the correct domain using the `AWS_LAMBDA_RUNTIME_API` env var.

remeh committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into remeh/serverless-logs

remeh committed:
serverless: re-use the daemon http server to receive the logs from AWS.

Instead of spawning a new HTTP server (using yet another tcp port), we are re-using
the already spawned HTTP server which is already used to communicate with the libraries.

It also means there is no more configuration field to change the HTTP. This may have
to be addressed if 8124 is not available for some reasons.

remeh committed:
dogstatsd: fix a unit test.

remeh committed:
Merge remote-tracking branch 'origin/master' into remeh/serverless-logs

remeh committed:
dogstatsd: fix a unit test.

remeh committed:
Merge remote-tracking branch 'origin/master' into remeh/serverless-logs

remeh committed:
tasks/dogstatsd: slighlty increase the maximum valid size for the dogstatsd binary.

prognant commented on pkg/config/log_format_serverless.go:
```diff
@@ -0,0 +1,21 @@
+	"github.com/cihub/seelog"
+)
+
+// buildCommonFormat returns the log common format seelog string
+func buildCommonFormat(loggerName LoggerName) string {
+	return fmt.Sprintf("%%Date(%s) | %s | %%LEVEL | %%Msg%%n", getLogDateFormat(), loggerName)
+}
```
I'm just curious, why do we remove contextual information from agent log when running in the serverless context ?
I guess here it's not a problem not to check for JMXFetch.

remeh replied:
We are mimicking the logs that users are used to have while running functions in AWS Lambda (through our clients libraries for instance).

prognant commented on pkg/config/log.go:
```diff
@@ -51,23 +51,6 @@ func createQuoteMsgFormatter(params string) seelog.FormatterFunc {
 
-// buildCommonFormat returns the log common format seelog string
-func buildCommonFormat(loggerName LoggerName) string {
-	if loggerName == "JMXFETCH" {
-		return `%Msg%n`
-	}
```
I think this should be included in `pkg/config/log_format.go`, for both `buildCommonFormat` and `buildJSONFormat`.

remeh replied:
Very nice catch, this must be a merge error or merge artifact.

prognant reviewed (commented):
For the logs agent part it looks good to me. I think it will works pretty well.

remeh committed:
misc: use the correct log-format for non-serverless logs.

This must be a merge error/artifact.

remeh committed:
Merge remote-tracking branch 'origin/master' into remeh/serverless-logs

prognant reviewed (approved):
LGTM, just a quick question about tagging, will be there any tag to specifically isolate one lambda execution stream ?

remeh commented:
@prognant thanks for the reviews... 🙇 

> will be there any tag to specifically isolate one lambda execution stream ?

You mean to identify that data sent by the Agent has been sent through a Lambda using the Extension? The `functionname` tag with the function name as the value is set on logs and on the enhanced metrics, but for now, not much more to differentiate logs emitted from a regular agent from logs emitted from the extension. Is this what you were wondering?

prognant commented:
> @prognant thanks for the reviews... 🙇
> 
> > will be there any tag to specifically isolate one lambda execution stream ?
> 
> You mean to identify that data sent by the Agent has been sent through a Lambda using the Extension? The `functionname` tag with the function name as the value is set on logs and on the enhanced metrics, but for now, not much more to differentiate logs emitted from a regular agent from logs emitted from the extension. Is this what you were wondering?

I was thinking about isolating/filtering logs (could be metrics) emitted by one single execution of a given lambda function. I.e. if a lambda function has a huge number of concurrent invocation, having a tag (like a unique uuid for each lambda execution) per lambda execution would be useful, WDYT ?

remeh commented:
> I was thinking about isolating/filtering logs (could be metrics) emitted by one single execution of a given lambda function. I.e. if a lambda function has a huge number of concurrent invocation, having a tag (like a unique uuid for each lambda execution) per lambda execution would be useful, WDYT ?

Oh sorry, I've read your question too fast! We already support that with the request ID which is set here: https://github.com/DataDog/datadog-agent/pull/6861/files#diff-353b87bf1595b02252db2d4fdae2de92f819a993f02f0d4a78e76ca174a96f51R82-R85
This request ID is then sent in the JSON and it is available in the Datadog app, for instance in the log explorer, or to filter by a given request id, etc. 👍
